src/Microsoft.ML.Recommender/SafeTrainingAndModelBuffer.cs

```diff
@@ -17,65 +17,44 @@ namespace Microsoft.ML.Runtime.Recommender.Internal
     /// </summary>
     internal sealed class SafeTrainingAndModelBuffer : IDisposable
     {
-        [StructLayout(LayoutKind.Explicit)]
+        [StructLayout(LayoutKind.Sequential)]
```

Comment:
This is not how we typically handle this on the .NET Core team.

Instead, what we typically do is define a static/stable binary interface from the C library, and P/Invoke into that. This involves doing conversion from native sized values like `int` into stable values - `int32_t` - which maps directly to `System.Int32`.

I suggest following the same pattern that we do elsewhere. It has treated us very well on the .NET Core team.

/cc @wschin

Reply:
@eerhardt, could you share some examples here?

Reply:
@wschin - checkout https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/interop-guidelines.md#unix-shims for the guidance we've been using on the .NET Core team. There are some guidelines that are corefx specific - like the naming guidelines (`SystemNative_`).

Here is an example of the `passwd` struct:

The definition in C++ code:

https://github.com/dotnet/corefx/blob/88a0580e61ea757db20347f114802e73375984dd/src/Native/Unix/System.Native/pal_uid.h#L11-L23

Notice that it uses static sized integers: `uint32_t` - it will always be an unsigned 32-bit integer.

Here is a C++ method that returns the structure:

https://github.com/dotnet/corefx/blob/a10890f4ffe0fadf090c922578ba0e606ebdd16c/src/Native/Unix/System.Native/pal_uid.c#L62-L77

And then the corresponding matching structure in C#:

https://github.com/dotnet/corefx/blob/a10890f4ffe0fadf090c922578ba0e606ebdd16c/src/Common/src/Interop/Unix/System.Native/Interop.GetPwUid.cs#L12-L23